### DISCUSSION TITLE
Mixing OpenMP with other threading paradigms

### DISCUSSION THREAD

**Discussion by gkaf89** [OpenMathLib/OpenBLAS#4332] [Q&A] [answered]
Regarding the compilation options, does it make sense to enable both `USE_OPENMP` and `USE_THREAD`?

In the file `driver/others/memory.c` I noticed that the use of some threading related code is always deactivated when OpenMP is used. For instance the directive 
```c
#if (defined(SMP) || defined(USE_LOCKING)) && !defined(USE_OPENMP)
...
#endif
```
is often used. Will this cause any issues?

**Comment by martin-frbg** [comment, marked as answer]
Yes this can/should both be enabled and USE_THREAD is on by default when you have more than one core in the build system. On many/most platforms, OpenMP is implemented on top of pthreads anyway. 
Only use USE_THREAD=0 (on its own) when you want an OpenBLAS that does not use any form of multithreading at all (which _can_ make sense if you have a big multithreaded program that makes many parallel calls to BLAS functions).
The directive you quoted is basically "if we are using multithreading, or expect to be called by multiple threads from outside,  and we do not have OpenMP manage the locking for thread-safe concurrent accesses"
